cmd/clusterctl/client/tree/discovery.go

```diff
@@ -105,25 +119,112 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 	controlPlaneMachines := selectControlPlaneMachines(machinesList)
 	for i := range controlPlaneMachines {
 		cp := controlPlaneMachines[i]
-		addMachineFunc(controlPLane, cp)
+		addMachineFunc(controlPlane, cp)
 	}
 
-	if len(machinesList.Items) == len(controlPlaneMachines) {
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(machinesList.Items) == len(controlPlaneMachines) && len(machinePoolList.Items) == 0 {
 		return tree, nil
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
 	tree.Add(cluster, workers)
 
+	err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	addMachinePoolsToObjectTree(cluster, workers, machinePoolList, tree, options)
+
+	// Handles orphan machines.
+	if len(machineMap) < len(machinesList.Items) {
+		other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+		tree.Add(workers, other)
+
+		for i := range machinesList.Items {
+			m := &machinesList.Items[i]
+			if _, ok := machineMap[m.Name]; ok {
+				continue
+			}
+			addMachineFunc(other, m)
+		}
+	}
+
+	return tree, nil
+}
+
+func constructPlaceholderObject(objectRef *corev1.ObjectReference) *unstructured.Unstructured {
+	placeholder := unstructured.Unstructured{}
+	placeholder.SetKind(objectRef.Kind)
+	placeholder.SetNamespace(objectRef.Namespace)
+	placeholder.SetName(objectRef.Name)
+	placeholder.SetAPIVersion(objectRef.APIVersion)
+	placeholder.SetUID(types.UID(objectRef.Kind + "/" + objectRef.Namespace + "/" + objectRef.Name))
+
+	return &placeholder
+}
+
+func addClusterResourceSetsToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, tree *ObjectTree) {
+	if resourceSetBinding, err := getResourceSetBindingInCluster(ctx, c, cluster.Namespace, cluster.Name); err == nil {
+		resourceSetGroup := VirtualObject(cluster.Namespace, "ClusterResourceSetGroup", "ClusterResourceSets")
+		tree.Add(cluster, resourceSetGroup)
+
+		owners := make(map[string]struct{})
+		for _, owner := range resourceSetBinding.GetOwnerReferences() {
+			if owner.Kind == "ClusterResourceSet" {
+				owners[owner.Name] = struct{}{}
+			}
+		}
+		for _, binding := range resourceSetBinding.Spec.Bindings {
+			resourceSetPlaceholder := constructPlaceholderObject(&corev1.ObjectReference{
+				Kind:       "ClusterResourceSet", // TODO: does this kind name exist anywhere as a const?
+				Namespace:  cluster.Namespace,
+				Name:       binding.ClusterResourceSetName,
+				APIVersion: addonsv1.GroupVersion.String(),
+			})
+			tree.Add(resourceSetGroup, resourceSetPlaceholder, ObjectMetaName("ClusterResourceSet"))
+			if _, ok := owners[binding.ClusterResourceSetName]; ok {
+				tree.Add(resourceSetPlaceholder, resourceSetBinding, ObjectMetaName("ClusterResourceSetBinding")) // Show binding under it's parent
+				owners = make(map[string]struct{})
+			}
```

Comment:
Is it required to show the binding under the cluster resource set when they are 1:1? 
If I got the goal right, what we care about is to show that a cluster is related to a ClusterResourceSet, and the fact that a binding existing is just an implementation detail